llvm/test/CodeGen/AMDGPU/llc-pipeline.ll

```diff
@@ -1170,8 +1166,7 @@
 ; GCN-O3-NEXT:    CallGraph Construction
 ; GCN-O3-NEXT:    Call Graph SCC Pass Manager
 ; GCN-O3-NEXT:      AMDGPU Annotate Kernel Features
-; GCN-O3-NEXT:      FunctionPass Manager
-; GCN-O3-NEXT:        AMDGPU Lower Kernel Arguments
+; GCN-O3-NEXT:    AMDGPU Lower Kernel Arguments
```

Comment:
Can this be a CGSCC pass instead?

Reply:
I've made this change. I'm worried that this leaves the SCC in an inconsonant state, although it works. It breaks if I try to remove any functions outside of finalization.

There are 3 other options. Was wondering if you have any opinion on the best one.
1. Module pass
2. Keep it a function pass and do all the preload stuff in the doInitialization callback
3. Make the preloading stuff a separate module pass and keep lowerkernelarguments a function pass
4. CGSCC pass

I'm leaning towards option 3 personally.

Reply:
wrt some of the (legacy) passes: it may not be an issue for now I'd be cautious adding module passes in the (Pre)CodeGen part as it pops off the CGSCC pass off the pass manager stack. From what I see in `llc-pipeline.ll`, it shouldn't be an issue but it's the reason why a few lines under where you add the pass in `AMDGPUTargetMachine` you see a sudden `addPass(new DummyCGSCCPass());` added: it pushes the CGSCC pass back on the pass manager stack.

Related to that: having the pass be a function pass will more than likely run the functions in CGSCC order anyways since it seems to inherit the CGSCC pass manager from the `AMDGPUAnnotateKernelFeatures` pass before it.

TL;DR: fine to add module pass in this case but may need a comment in `AMDGPUTargetMachine` mentioning that it pops off the CGSCC pass manager from the PM stack for future explorers of said code.

Reply:
Thanks @JanekvO I'll add a comment to that effect.

@arsenm to future proof for NPM I think our only options are making AMDGPULowerKernelArguments a module pass or factoring out the preload stuff and creating a new module pass. What do you think is better?

Reply:
I thought you were moving the preload logic to the attributor? This probably belongs in the attributor, or a separate CGSCC/CGSCC-in-module pass.

This does not feel like it belongs in AMDGPULowerKernelArguments. AMDGPULowerKernelArguments is at least a 50% hack due to clang emitting sub-optimal kernel signatures to begin with (which #115821 will soon give us a path forward to fixing). The other piece is it enables vectorization of kernel argument loads, which will not go away. In either case, its only job is to rewrite argument references in terms of a base pointer. It shouldn't need to consider the use context like preload should